Migration docs for DCIM Filterset Changes (#2917)
nautobot/nautobot: nautobot/docs/release-notes/version-2.0.md

`/dcim/regions/?parent=<slug>`

Now in v2.x, you are able to filter those `Regions` by slugs or UUID primary keys:

`/dcim/regions/?parent=<slug>` or `/dcim/regions/?parent=<uuid>`

Check out the specific changes documented in the table at [Updating from Nautobot 1.5 to 2.0](../installation/upgrading.md/#enhanced-filter-fields)

#### Corrected Filter Fields ([#2804](https://github.com/nautobot/nautobot/pull/2804))

There were also instances where a foreign-key related field (e.g. `console_ports`) was incorrectly mapped to a boolean membership filter (e.g. `has_console_ports`), making it impossible to filter based on specific values of the foreign key:

For example in v1.x:

`/dcim/devices/?console_ports=True` and `/dcim/devices/?has_console_ports=True` are functionally the same and this behavior is **incorrect**.

This has been addressed in v2.x as follows:

`console_ports` and similar filters are taking foreign key UUIDs as input values and can be used in this format: `/dcim/devices/?console_ports=<uuid>` whereas `has_console_ports` and similar filters remain the same.

> Check out the specific changes documented in the table at [Updating from Nautobot 1.5 to 2.0](../installation/upgrading.md/#corrected-filter-fields)

Reviewer: ```suggestion
Check out the specific changes documented in the table at [Updating from Nautobot 1.5 to 2.0](../installation/upgrading.md#corrected-filter-fields)
```

Double check these links: https://docs-landing-page--2917.org.readthedocs.build/projects/core/en/stable/installation/release-notes/version-2.0.md/#enhanced-filter-fields-2804

### Removed

#### Removed Redundant Filter Fields ([#2804](https://github.com/nautobot/nautobot/pull/2804))

> As a part of breaking changes made in v2.X, shadowed filter/filterset fields are being removed in the dcim app:

Reviewer 2: ```suggestion
As a part of breaking changes made in v2.X, shadowed filter/filterset fields are being removed in the DCIM app:
```

### Removed

> #### Removed DCIM Shadowed Filter Fields ([#2804](https://github.com/nautobot/nautobot/pull/2804))

Reviewer 2: Perhaps:

```suggestion
#### Removed Redundant Filter Fields ([#2804](https://github.com/nautobot/nautobot/pull/2804))
```

As a part of breaking changes made in v2.0, shadowed filter/filterset fields are being removed in the dcim app:

> Currently for foreign-key related fields, **on existing core models in the v1.3 release train**:

Reviewer 2: Why 1.3 specifically?

    - The field is shadowed for the purpose of replacing the PK filter with a lookup-based on a more human-readable value (typically `slug`, if available).
    - A PK-based filter is available as well, generally with a name suffixed by `_id`

For example:

```python
    site = django_filters.ModelMultipleChoiceFilter(
        field_name="site__slug",
        queryset=Site.objects.all(),
        to_field_name="slug",
        label="Site (slug)",
    )
    site_id = django_filters.ModelMultipleChoiceFilter(
        queryset=Site.objects.all(),
        label="Site (ID)",
    )

> ```

Reviewer 2: Instead of the source code, which is not perhaps useful to the end user, it might be more helpful to include examples of the filtering, e.g. `/dcim/devices/?site_id=<uuid>` in 1.x versus `/dcim/devices/?site=<uuid>` in 2.x.

> Now these two fields will be replaced by a single field using `NaturalKeyOrPKMultipleChoiceFilter` introduced in v1.4:

Reviewer 2: Similarly, it might be more helpful to say something like:

```suggestion
Now these two filter fields will be replaced by a single filter field that can support both slugs and UUID primary keys as inputs:
```

```python
    from nautobot.utilities.filters import NaturalKeyOrPKMultipleChoiceFilter
    site = NaturalKeyOrPKMultipleChoiceFilter(
        queryset=Site.objects.all(),
        label="Site (slug or ID)",
    )
```

Below is a table documenting all such changes and where they occurred.

| Filterset                      | Old Filter Field   | Changes                                                                      |
|--------------------------------|--------------------|------------------------------------------------------------------------------|

> | LocatableModelFilterSetMixin   | region_id          | Deleted                                                                      |

Reviewer 2: Referencing internal classes isn't very helpful to the end user. Would be better to enumerate the UI and REST API endpoints that are impacted by these changes.

|                                | site_id            | Deleted                                                                      |
| RegionFilterSet                | parent_id          | Deleted                                                                      |

> |                                | parent             | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |

Reviewer 2: ```suggestion
|                                | parent             | Enhanced to support primary key UUIDs in addition to slugs                        |
```

Similar suggestion throughout.

> |                                | parent             | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |

Reviewer 2: Listing all of this under a "Removed" heading isn't ideal either. Consider either moving this to "Changed" or splitting the "removed" filter fields separate from the "changed/enhanced" filter fields?

| RackGroupFilterSet             | parent_id          | Deleted                                                                      |
|                                | parent             | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |
| RackFilterSet                  | group_id           | Deleted                                                                      |
|                                | role_id            | Deleted                                                                      |
|                                | role               | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |

> |                                | serial             | Changed to use `MultiValueCharFilter`                                        |

Reviewer 2: Perhaps something like:
```suggestion
|                                | serial             | Enhanced to permit filtering on multiple values                                        |
```

|                                | tenant_id          | Deleted                                                                      |
|                                | tenant             | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |
| PowerPanelFilterSet            | rack_group_id      | Deleted                                                                      |
| PowerFeedFilterSet             | region_id          | Deleted                                                                      |
|                                | site_id            | Deleted                                                                      |
|                                | site               | Changed to use `NaturalKeyOrPKMultipleChoiceFilter`                          |
|                                | power_panel_id     | Deleted                                                                      |
|                                | rack_id            | Deleted                                                                      |

> There are also instances where a foreign-key related field is incorrectly mapped to a boolean membership filter.

Reviewer 2: ```suggestion
There were also instances where a foreign-key related field was incorrectly mapped to a boolean membership filter, making it impossible to filter based on specific values of the foreign key. This has been addressed as follows ...
```

For example:

```python
    has_power_outlets = RelatedMembershipBooleanFilter(
            field_name="poweroutlets",
            label="Has power outlets",
        )
    power_outlets = has_power_outlets
```

Now we change the foregin-key related field to be its own PK based filter:

```python
    has_power_outlets = RelatedMembershipBooleanFilter(
            field_name="poweroutlets",
            label="Has power outlets",
        )
    power_outlets = django_filters.ModelMultipleChoiceFilter(
        queryset=PowerOutlet.objects.all(),
        field_name="poweroutlets",
        label="Power Outlets",
    )

> ```

Reviewer 2: here too something like:

before: `/dcim/devices/?power_outlets=true`, after: `/dcim/devices/?power_outlets=<uuid>` or `/dcim/devices/?has_power_outlets=true`

| DeviceFilterSet | console_ports        | Changed from mapping to `has_console_ports` to `ModelMultipleChoiceFilter`        |
|                 | console_server_ports | Changed from mapping to `has_console_server_ports` to `ModelMultipleChoiceFilter` |
|                 | power_ports          | Changed from mapping to `has_power_ports` to `ModelMultipleChoiceFilter`          |
|                 | power_outlets        | Changed from mapping to `has_power_outlets` to `ModelMultipleChoiceFilter`        |
|                 | interfaces           | Changed from mapping to `has_interfaces` to `ModelMultipleChoiceFilter`           |
|                 | front_ports          | Changed from mapping to `has_front_ports` to `ModelMultipleChoiceFilter`          |
|                 | rear_ports           | Changed from mapping to `has_rear_ports` to `ModelMultipleChoiceFilter`           |
|                 | device_bays          | Changed from mapping to `has_device_bays` to `ModelMultipleChoiceFilter`          |
|                 | pass_through_ports   | Deleted in exchange for in exchange for `has_(front/rear)_ports`                  |

!!! important

>     `NaturalKeyOrPKMultipleChoiceFilter` is not filtering on true Natural Keys in the case where its `to_field_name` argument is mapped to the `name` attribute or a single field that is not enough to identify an unique model instance by itself. This is technically a bug and is tracked here <https://github.com/nautobot/nautobot/issues/2875> and should be resolved after concluding this Epic <https://github.com/nautobot/nautobot/issues/1574>.

Reviewer 2: A concrete example here might be more helpful to the end user than the technical details of how it's implemented.

Author: I feel like this portion might be too technical for an end user facing documentation so I am going to delete this. 
We also have the #TODO comments in this [PR](https://github.com/nautobot/nautobot/pull/2882/) to track this issue internally.
